pkg/kubelet/cm/container_manager_linux.go

```diff
@@ -342,6 +345,137 @@ func setupKernelTunables(option KernelTunableBehavior) error {
+	// on the cgroup and prevent further growth.
+	glog.Warningf("[Container Manager] %s QoS memory usage is already above desired limit", qosClass)
+	stats, err := cgroupManager.Get(cgroupConfig.Name)
+	if err != nil {
+		return err
+	}
+	usageLimit := stats.MemoryStats.Usage + memoryUsageMargin
+	cgroupConfig.ResourceParameters.Memory = &usageLimit
+	glog.V(2).Infof("[Container Manager] Setting %s QoS memory limit to %d bytes", qosClass, usageLimit)
+	return cgroupManager.Update(&cgroupConfig)
+}
+
+// syncQOSMemoryLimts sums the memory limits of all pods in a QOS class,
+// calculate QOS class memory limits, and set those limits in the QOS cgroups
+func setQOSMemoryReserveLimit(cm *containerManagerImpl, reservePercent int64) error {
+	qosMemoryLimits := map[v1.PodQOSClass]int64{
+		v1.PodQOSGuaranteed: 0,
+		v1.PodQOSBurstable:  0,
+	}
+
+	// Sum the pod limits for pods in each QOS class
+	pods := cm.getPodsFunc()
+	for _, pod := range pods {
+		podMemoryLimit := int64(0)
+		qosClass := qos.GetPodQOS(pod)
+		if qosClass == v1.PodQOSBestEffort {
+			// limits are not set for Best Effort pods
+			continue
+		}
+		for _, container := range pod.Spec.Containers {
+			podMemoryLimit += container.Resources.Limits.Memory().Value()
+		}
+		qosMemoryLimits[qosClass] += podMemoryLimit
+	}
+
+	allocatable := cm.nodeInfo.Status.Allocatable.Memory().Value()
+	if allocatable == 0 {
+		return fmt.Errorf("memory capacity reported as 0, might be in standalone mode")
+	}
+
+	for qos, limits := range qosMemoryLimits {
+		glog.V(2).Infof("[Container Manager] %s pod limits total %d bytes (reserve %d%%)", qos, limits, reservePercent)
+	}
+
+	// Calculate QOS memory limits
+	burstableLimit := allocatable - (qosMemoryLimits[v1.PodQOSGuaranteed] * reservePercent / 100)
+	if burstableLimit < minimumLimit {
+		burstableLimit = minimumLimit
+	}
+	bestEffortLimit := burstableLimit - (qosMemoryLimits[v1.PodQOSBurstable] * reservePercent / 100)
+	if bestEffortLimit < minimumLimit {
+		bestEffortLimit = minimumLimit
+	}
+
+	// Update QOS class cgroups with calculated memory limits
+	subsystems, err := GetCgroupSubsystems()
+	if err != nil {
+		return err
+	}
+
+	cgroupManager := NewCgroupManager(subsystems, cm.CgroupDriver)
+	err = setQOSCgroupMemoryLimit(cgroupManager, v1.PodQOSBurstable, burstableLimit)
+	if err != nil {
+		return err
+	}
+	err = setQOSCgroupMemoryLimit(cgroupManager, v1.PodQOSBestEffort, bestEffortLimit)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type qosReserveLimitHandler func(*containerManagerImpl, int64) error
+
+var resourceQOSReserveLimitHandlers = map[string]qosReserveLimitHandler{
+	"memory": setQOSMemoryReserveLimit,
+}
+
+// If usage is already above the limit, set the limit to the usage plus this margin
+// This increases the likelyhood that the limit can be set successfully and gives the
+// cgroup a little forgiveness so it doesn't OOM kill in response to a little growth.
+const memoryUsageMargin = 10 * 1024 * 1024 // 10MB
```

Comment:
if you do mean this, i think that is a good (i.e. safer) idea.  soft limit typically don't get the OOM killer involved.

i think it would be a simple change, setting `MemoryReservation` rather than `Memory` in the `libcontainerconfigs.Resources`